CC0120: Suggest default for switch statements (C#)

A lot of bugs are caused by non-exhaustive matching in switch statements.
If there is no default case one should be added that throws an exception.

Before:

``` csharp
switch (a)
{
    case "":
        break;
}
```

After:

``` csharp
switch (a)
{
    case "":
        break;
    default:
        throw new Exception("Unexpected Case");
}
```

Should not raise for `boolean` if `true` and `false` are already handled:

``` csharp
var f = true;
switch (f)
{
    case true:
        break;
    case false:
        break;
    default:
        break;
}
```

Should not raise a diagnostic if [CS0151](https://msdn.microsoft.com/en-us/library/dh0x1stb.aspx)(C# A switch expression or case label must be a bool, char, string, integral, enum, or corresponding nullable type) is already raised in the switch.

Diagnostic Id: `CC0120`
Category: `Design`
Severity: `Warning`

Same as #767, but for C#.

@cezarant is working on it.

Hi Devs! 
Built the analyzer and codeFixer for this issue. The id is SwicthWithoutDefault. Also built four tests, for the following:  

1- Using a Boolean variable 
2- With the default clause already inserted 
3- With the compilation error CS0151
4- Without the Default clause

I merge locally and I'm ready to do the pull request. 
Are there any step that I can perform the pull request?

--- a/src/CSharp/CodeCracker/CodeCracker.csproj
+++ b/src/CSharp/CodeCracker/CodeCracker.csproj
@@ -46,6 +46,8 @@
     <Compile Include="Design\InconsistentAccessibility\InconsistentAccessibilityInPropertyType.cs" />
     <Compile Include="Design\MakeMethodStaticAnalyzer.cs" />
     <Compile Include="Design\MakeMethodStaticCodeFixProvider.cs" />
+    <Compile Include="Design\SwitchWithoutDefaultAnalyzer.cs" />
+    <Compile Include="Design\SwitchWithoutDefaultCodeFixProvider.cs" />
     <Compile Include="Extensions\CSharpAnalyzerExtensions.cs" />
     <Compile Include="Extensions\CSharpGeneratedCodeAnalysisExtensions.cs" />
     <Compile Include="Extensions\InitializerState.cs" />

--- a/src/CSharp/CodeCracker/Design/SwitchWithoutDefaultAnalyzer.cs
+++ b/src/CSharp/CodeCracker/Design/SwitchWithoutDefaultAnalyzer.cs
@@ -1,0 +1,66 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.Diagnostics;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using System.Collections.Immutable;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+
+
+namespace CodeCracker.CSharp.Design
+{
+    [DiagnosticAnalyzer(LanguageNames.CSharp)]
+    public class SwitchWithoutDefaultAnalyzer : DiagnosticAnalyzer
+    {
+        internal const string Title = "Your Switch maybe include default clause";
+        internal const string MessageFormat = "{0}";
+        internal const string Category = SupportedCategories.Design;
+        internal static readonly DiagnosticDescriptor Rule = new DiagnosticDescriptor(
+            DiagnosticId.SwitchCaseWithoutDefault.ToDiagnosticId(),
+            Title,
+            MessageFormat,
+            Category,
+            DiagnosticSeverity.Warning,
+            isEnabledByDefault: true,
+            helpLinkUri: HelpLink.ForDiagnostic(DiagnosticId.SwitchCaseWithoutDefault));
+
+        public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics => ImmutableArray.Create(Rule);
+
+        public override void Initialize(AnalysisContext context) => context.RegisterSyntaxNodeAction(Analyzer, SyntaxKind.SwitchStatement);
+
+        private static void Analyzer(SyntaxNodeAnalysisContext context)
+        {
+            if (context.IsGenerated()) return;
+            // checks if the compile error CS0151 it has fired.If Yes, don't report the diagnostic
+            var diagnostics = from dia in context.SemanticModel.GetDiagnostics()
+                              where dia.Id == "CS0151"
+                              select dia;
+            if (diagnostics.Any()) return;
+            if (context.Node.IsKind(SyntaxKind.SwitchStatement))
+            {
+                var switchStatementToAnalyse = (SwitchStatementSyntax)context.Node;
+                if (switchStatementToAnalyse.DescendantNodes().Where(n => n.Kind() == SyntaxKind.DefaultSwitchLabel).ToList().Count == 0)
+                {
+                    var hasInitializer = from nodes in switchStatementToAnalyse.DescendantNodes()
+                                         where nodes.Kind() == SyntaxKind.IdentifierName
+                                         select nodes;
+                    if (!hasInitializer.Any())
+                        return;
+                    var hasTrueExpression = from nodes in switchStatementToAnalyse.DescendantNodes()
+                                            where nodes.Kind() == SyntaxKind.FalseLiteralExpression
+                                            select nodes;
+                    var hasfalseExpression = from nodes in switchStatementToAnalyse.DescendantNodes()
+                                             where nodes.Kind() == SyntaxKind.TrueLiteralExpression
+                                             select nodes;
+                    if ((hasfalseExpression.Any()) && (hasTrueExpression.Any()))
+                        return;
+                    var diagnostic = Diagnostic.Create(Rule, switchStatementToAnalyse.GetLocation(), "Consider put an default clause in Switch.");
+                    context.ReportDiagnostic(diagnostic);
+                }
+            }
+        }
+    }
+}

--- a/src/CSharp/CodeCracker/Design/SwitchWithoutDefaultCodeFixProvider.cs
+++ b/src/CSharp/CodeCracker/Design/SwitchWithoutDefaultCodeFixProvider.cs
@@ -1,0 +1,121 @@
+﻿using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CodeActions;
+using Microsoft.CodeAnalysis.CodeFixes;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Formatting;
+using System.Collections.Generic;
+using System.Collections.Immutable;
+using System.Composition;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+namespace CodeCracker.CSharp.Design
+{
+    [ExportCodeFixProvider(LanguageNames.CSharp, Name = nameof(SwitchWithoutDefaultCodeFixProvider)), Shared]
+    public class SwitchWithoutDefaultCodeFixProvider : CodeFixProvider
+    {
+        private const string EmptyString = "\"\"";
+        private const string BreakString = "\n\t\t\t\t\tbreak;";
+        private const string ThrowString = "throw new Exception(\"Unexpected Case\");";
+        public sealed override ImmutableArray<string> FixableDiagnosticIds =>
+            ImmutableArray.Create(DiagnosticId.SwitchCaseWithoutDefault.ToDiagnosticId());
+        public sealed override FixAllProvider GetFixAllProvider() => WellKnownFixAllProviders.BatchFixer;
+
+        public sealed override Task RegisterCodeFixesAsync(CodeFixContext context)
+        {
+            var diagnostic = context.Diagnostics.First();
+            context.RegisterCodeFix(CodeAction.Create("Add a default clause", c => SwitchWithoutDefaultAsync(context.Document, diagnostic, c),
+                                    nameof(SwitchWithoutDefaultCodeFixProvider)), diagnostic);
+            return Task.FromResult(0);
+        }
+
+        private async static Task<Document> SwitchWithoutDefaultAsync(Document document, Diagnostic diagnostic, CancellationToken cancellationToken)
+        {
+            var root = await document.GetSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var switchCaseLabel = new SyntaxList<SwitchLabelSyntax>();
+            var kindOfVariable = string.Empty;
+            var idForVariable = string.Empty;
+            var breakStatement = new SyntaxList<StatementSyntax>();
+            var diagnosticSpan = diagnostic.Location.SourceSpan;
+            var newSections = new List<SwitchSectionSyntax>();
+            var switchCaseStatement = root.FindToken(diagnosticSpan.Start).Parent.AncestorsAndSelf().OfType<SwitchStatementSyntax>().First();
+            idForVariable = ((IdentifierNameSyntax)switchCaseStatement.ChildNodes().ToList().First(n => n.Kind() == SyntaxKind.IdentifierName)).Identifier.ValueText;
+            var parametersOfMethod = from init in root.DescendantNodesAndSelf()
+                                     where init.Kind() == SyntaxKind.Parameter
+                                     select init;
+            // Verify if the variable of Switch as a same of Method Parameter 
+            foreach (var parameter in parametersOfMethod)
+                if (idForVariable == ((ParameterSyntax)parameter).Identifier.ValueText)
+                    kindOfVariable = ((ParameterSyntax)parameter).Type.ToString();
+
+            if (kindOfVariable == string.Empty)
+            {
+                var statements = ((BlockSyntax)switchCaseStatement.Parent).Statements;
+                var inicializerOfSwitch = from init in statements
+                                          where init.Kind() == SyntaxKind.LocalDeclarationStatement
+                                          select init;
+                if (inicializerOfSwitch.Any())
+                {
+                    var local = (LocalDeclarationStatementSyntax)inicializerOfSwitch.First();
+                    var switchCaseVariable = ((local).Declaration).Variables[0];
+                    kindOfVariable = ((LiteralExpressionSyntax)(switchCaseVariable.Initializer.Value)).Kind().ToString();
+                }
+            }
+
+            if ((kindOfVariable.Equals("FalseLiteralExpression")) || (kindOfVariable.Equals("TrueLiteralExpression"))
+                || (kindOfVariable.Equals("bool")))
+            {
+                var oldSections = switchCaseStatement.Sections.ToList();
+                var type = string.Empty;
+                foreach (var sec in oldSections)
+                {
+                    newSections.Add(sec);
+                    type = (((CaseSwitchLabelSyntax)((SwitchSectionSyntax)sec).ChildNodes().ToList().First()).Value).GetFirstToken().Text;
+                }
+                var againstType = string.Empty;
+                if (type.Equals("true")) againstType = "false"; else againstType = "true";
+
+                newSections.Add(SyntaxFactory.SwitchSection().WithLabels(
+                            switchCaseLabel.Add(SyntaxFactory.CaseSwitchLabel(SyntaxFactory.ParseExpression(againstType)))).
+                               WithStatements(breakStatement.Add(SyntaxFactory.ParseStatement(BreakString))));
+            }
+            else
+            {
+                var oldSections = switchCaseStatement.Sections.ToList();
+                foreach (var sec in oldSections)
+                    newSections.Add(sec);
+
+                breakStatement = breakStatement.Add(SyntaxFactory.ParseStatement(BreakString));
+                newSections.Add(CreateSection(SyntaxFactory.DefaultSwitchLabel(), SyntaxFactory.ParseStatement(ThrowString)));
+            }
+            var switchExpression = SyntaxFactory.ParseExpression(idForVariable);
+            var newsSwitchCaseStatement = SyntaxFactory.SwitchStatement(switchExpression).
+                            WithSections(new SyntaxList<SwitchSectionSyntax>().AddRange(newSections));
+            var newRoot = root.ReplaceNode(switchCaseStatement, newsSwitchCaseStatement);
+            var newDocument = document.WithSyntaxRoot(newRoot);
+            return newDocument;
+        }
+        static SwitchSectionSyntax CreateSection(SwitchLabelSyntax label, StatementSyntax statement)
+        {
+            var labels = new SyntaxList<SwitchLabelSyntax>();
+            labels = labels.Add(label);
+            return SyntaxFactory.SwitchSection(labels, CreateSectionStatements(statement));
+        }
+
+        static SyntaxList<StatementSyntax> CreateSectionStatements(StatementSyntax source)
+        {
+            var result = new SyntaxList<StatementSyntax>();
+            if (source is BlockSyntax)
+            {
+                var block = source as BlockSyntax;
+                result = result.AddRange(block.Statements);
+            }
+            else
+            {
+                result = result.Add(source);
+            }
+            return result;
+        }
+    }
+}

--- a/src/Common/CodeCracker.Common/DiagnosticId.cs
+++ b/src/Common/CodeCracker.Common/DiagnosticId.cs
@@ -81,5 +81,6 @@
         StringFormatArgs_ExtraArgs = 111,
         AlwaysUseVarOnPrimitives = 105,
         UnnecessaryToStringInStringConcatenation = 118,
+        SwitchCaseWithoutDefault = 120,
     }
 }

--- a/test/CSharp/CodeCracker.Test/CodeCracker.Test.csproj
+++ b/test/CSharp/CodeCracker.Test/CodeCracker.Test.csproj
@@ -129,6 +129,7 @@
     <Compile Include="Design\InconsistentAccessibilityTests.MethodIndexerParameter.cs" />
     <Compile Include="Design\InconsistentAccessibilityTests.MethodIndexerReturnType.cs" />
     <Compile Include="Design\MakeMethodStaticTests.cs" />
+    <Compile Include="Design\SwicthWithoutDefaultTests.cs" />
     <Compile Include="GeneratedCodeAnalysisExtensionsTests.cs" />
     <Compile Include="GlobalSuppressions.cs" />
     <Compile Include="Performance\UseStaticRegexIsMatchTests.cs" />

--- a/test/CSharp/CodeCracker.Test/Design/SwicthWithoutDefaultTests.cs
+++ b/test/CSharp/CodeCracker.Test/Design/SwicthWithoutDefaultTests.cs
@@ -1,0 +1,353 @@
+﻿using CodeCracker.CSharp.Design;
+using System.Threading.Tasks;
+using Xunit;
+
+namespace CodeCracker.Test.CSharp.Design
+{
+    public class SwicthWithoutDefaultTests : CodeFixVerifier<SwitchWithoutDefaultAnalyzer, SwitchWithoutDefaultCodeFixProvider>
+    {
+        [Fact]
+        public async Task SwithWithoutDefaultAnalyserString()
+        {
+            const string source = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+                                              static void Main()
+                                              {
+var a = """";
+            switch (a)
+            {
+                case """":
+                    break;                
+            }
+                                              }
+                                        }
+                                   }";
+
+            const string fixtest = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+                                              static void Main()
+                                              {
+var a = """";
+            switch (a)
+            {
+                case """":
+                    break;
+                default:
+                    throw new Exception(""Unexpected Case"");
+            }
+    }
+                                        }
+                                   }";
+            await VerifyCSharpFixAsync(source, fixtest, 0, allowNewCompilerDiagnostics: true);
+        }
+
+        [Fact]
+        public async Task SwithWithoutDefaultAnalyserBool()
+        {
+            const string source = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+                                              static void Main()
+                                              {
+                                                var s = true;
+                                                switch (s)
+                                                {
+                                                    case false:
+                                                    break;                
+                                                }
+                                        }
+                                   }";
+
+            const string fixtest = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+                                              static void Main()
+                                              {
+                                                var s = true;
+                                                switch (s)
+                                                {
+                                                    case false:
+                                                        break;
+                                                    case true:
+                                                        break;
+                                                }
+                                        }                                        
+                                   }";
+            await VerifyCSharpFixAsync(source, fixtest, 0, allowNewCompilerDiagnostics: true);
+        }
+
+        [Fact]
+        public async Task SwithWithoutDefaultAnalyserInt()
+        {
+            const string source = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+                                              static void Main()
+                                              {
+var s = 10;
+            switch (s)
+            {
+                case 10:
+                    break;                
+            }
+
+}
+                                        }
+                                   }";
+
+            const string fixtest = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+                                              static void Main()
+                                              {
+var s = 10;
+            switch (s)
+            {
+                case 10:
+                    break;
+                default:
+                    throw new Exception(""Unexpected Case"");
+            }
+    }
+}
+                                   }";
+            await VerifyCSharpFixAsync(source, fixtest, 0, allowNewCompilerDiagnostics: true);
+
+        }
+
+        public async Task SwithWithoutDefaultAnalyserBoolMethod()
+        {
+            const string source = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+                                              void Bar(bool myBool)
+        {
+            switch (myBool)
+            {
+                case false:
+                    break;                
+            }
+        }
+                                        }
+                                   }";
+
+            const string fixtest = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+                                            void Bar(bool myBool)
+        {
+            switch (myBool)
+            {
+                case false:
+                    break;
+                case true:
+                    break;
+            }
+        }
+    
+                                        }
+                                   }";
+            await VerifyCSharpFixAsync(source, fixtest, 0, allowNewCompilerDiagnostics: true);
+        }
+
+        public async Task SwithWithoutDefaultAnalyserIntMethod()
+        {
+            const string source = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+ void Bar3(int a)
+        {            
+            switch (a)
+            {
+                case 10:
+                    break;                
+            }
+        }     
+                                        }
+                                   }";
+
+            const string fixtest = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+         void Bar3(int a)
+        {
+            switch (a)
+            {
+                case 10:
+                    break;
+                default:
+                    throw new Exception(""Unexpected Case"");
+            }
+       }
+}
+                                   }";
+            await VerifyCSharpFixAsync(source, fixtest, 0, allowNewCompilerDiagnostics: true);
+        }
+
+        public async Task SwithWithoutDefaultAnalyserIntMethodTwoParams()
+        {
+            const string source = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+void Bar4(int a, int b)
+        {
+            switch (a)
+            {
+                case 10:
+                    break;                
+            }
+        }
+                                        }
+                                   }";
+
+            const string fixtest = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {        
+void Bar4(int a, int b)
+        {
+            switch (a)
+            {
+                case 10:
+                    break;
+                default:
+                    throw new Exception(""Unexpected Case"");
+            }
+    }
+}
+                                   }";
+            await VerifyCSharpFixAsync(source, fixtest, 0, allowNewCompilerDiagnostics: true);
+        }
+
+        [Fact]
+        public async Task SwithWithoutDefaultAnalyserIntMethodStatic()
+        {
+            const string source = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+static void Bar5(int a)
+        {
+            switch (a)
+            {
+                case 10:
+                    break;                
+            }
+        }
+
+                                        }
+                                   }";
+
+            const string fixtest = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {        
+static void Bar5(int a)
+        {
+            switch (a)
+            {
+                case 10:
+                    break;
+                default:
+                    throw new Exception(""Unexpected Case"");
+            }
+    }
+
+}
+                                   }";
+            await VerifyCSharpFixAsync(source, fixtest, 0, allowNewCompilerDiagnostics: true);
+        }
+
+        [Fact]
+        public async Task SwithWithoutDefaultAnalyseCS0151Exception()
+        {
+            const string source = @"static void M() { }
+        static void Main()
+        {
+            
+            switch (M()) // CS0151
+            {
+                default:
+                    break;
+            }
+        }";
+            await VerifyCSharpHasNoDiagnosticsAsync(source);
+        }
+
+
+        [Fact]
+        public async Task SwithWithoutDefaultAnalyserNoDiagnostic()
+        {
+            const string source = @"using System;namespace 
+                                    ConsoleApplication1 
+                                    {
+                                        class TypeName
+                                        {
+                                              static void Main()
+                                              {
+                                                var s = "";
+                                                 switch (s)
+                                                 {
+                                                  case "":{break;}
+                                                  default:{break;}
+                                                 }
+                                        }
+                                      }
+                                   }";
+            await VerifyCSharpHasNoDiagnosticsAsync(source);
+        }
+
+        [Fact]
+        public async Task SwitchWithoutDefaultAnalyzerNoDiagnosticWithCompileError()
+        {
+            const string source = @"using System;
+                                    namespace ConsoleApplication1
+                                    { 
+                                        ConsoleApplication1 
+                                        {
+                                            class Teste { }
+                                            class Program
+                                            {
+                                                static void Main(string[] args)
+                                                {
+                                                    Teste vo_teste = new Teste(); 
+                                                    switch (vo_teste)
+                                                    {
+                                                        case "":break; 
+                                                        default:break;
+                                                    }
+                                                }
+                                            } 
+                                        }                                     
+                                   }";
+            await VerifyCSharpHasNoDiagnosticsAsync(source);
+        }
+    }
+}

